Make Apache control script and binary paths configurable on command line

joohoi opened the pull request:
This PR adds two new command line parameters, `--apache-ctlpath` and `--apache-binpath` both of which are used to construct commands that we shell out for. 

The way that we previously fetched values either from Certbot configuration object or the dictionary of distribution based constants is now also unified, and the active options are parsed in `prepare()` to make it easier to override needed values for the distributions needing this behavior.

Fixes: #5338

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -94,26 +95,49 @@ class ApacheConfigurator(augeas_configurator.AugeasConfigurator):
         vhost_files="*",
         logs_root="/var/log/apache2",
+        ctlpath="apache2ctl",
+        binpath="apache2",
```
I thought that it might clarify the need of full paths for users that are running custom Apache installations (the main demographic for users of these flags), as these might not have the `$PATH` correctly set.

joohoi replied:
I do agree, this is now changed. Shorter options are always nicer, especially if you need to define multiple of them.

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -94,26 +95,49 @@ class ApacheConfigurator(augeas_configurator.AugeasConfigurator):
+        """
+        opts = ["enmod", "dismod", "le_vhost_ext", "server_root", "vhost_root",
+                "logs_root", "challenge_location", "handle_modules", "handle_sites",
+                "ctlpath", "binpath"]
```
I don't think it's easily achievable, as `add_parser_arguments` is a classmethod, and the changes made to it do not make it to the actual instance we'd be using when doing `prepare()`. One way to avoid this would be to add a `list` of `dict`s, each of which describe an command line argument as a class constant, but I feel that this would obscure the behavior more. For now, I added a comment reminding of changing the list in `_prepare_options()` as well. Do you feel that this would be enough?

bmw replied:
Why doesn't my proposal work? I suggested we make it a `set` to avoid issues if `add_parser_arguments` is called more than once and suggested making the `set` a global or class variable to work around `add_parser_arguments` being a class method. Am I missing something?

I'm still interested in an approach like this as it seems relatively simple and makes it easy to do things right in the future, but adding a comment is fine for now if you prefer.

joohoi replied:
It's because `add_parser_arguments` are called on the class itself, and not an instance of it, so it doesn't allow us to store the added values to itself. Global variable could work, but I'm kind of thinking global variables largely as antipattern, my thinking might be flawed though.

bmw replied:
I was specifically referring to a class variable like `OS_DEFAULTS` instead an instance variable to work around this.

It's all a tradeoff though. Variables with a large scope can cause problems but so can trying to remember to properly make changes to multiple places in the future. A comment is fine for now.

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -94,26 +95,49 @@ class ApacheConfigurator(augeas_configurator.AugeasConfigurator):
+                "ctlpath", "binpath"]
+        for o in opts:
+            # Config options use dashes instead of underscores
+            if self.conf(o.replace("_", "-")):
```
Oh yeah, this was a legit bug. Thanks!

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -132,12 +156,16 @@ def add_parser_arguments(cls, add):
+        add("ctlpath", default=cls.OS_DEFAULTS["ctlpath"],
+            help="Path to Apache control script")
+        add("binpath", default=cls.OS_DEFAULTS["binpath"],
+            help="Path to Apache binary or a script passing arguments to it.")
```
Agreed, all of the help texts for Apache are now fixed.

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -94,26 +95,49 @@ class ApacheConfigurator(augeas_configurator.AugeasConfigurator):
+
+        # Special cases
+        self.options["apache_cmd"] = self.option("ctlpath")
+        self.options["version_cmd"][0] = self.option("binpath")
```
I did change it for consistency, the change was on purpose.

bmw replied:
OK just making sure.

Now that we're using a new executable, should we add a [check in prepare](https://github.com/certbot/certbot/blob/cdc333491be3cc48dd17c5191f815859035bfdc7/certbot-apache/certbot_apache/configurator.py#L205) that the executable exists (and call `path_surgery` to fix up the user's paths if not)?

Also, did you check that this works on all the different OSes being modified here?

joohoi replied:
The check is moved to a helper method and both of the executables are now checked.

The question of all different OSes is a harder one, because of the implementation in `configurator.py` is a catch-all for all distros that don't have a matching override.

Now when rethinking this decision it might be best to use the old value to ensure compatibility.

bmw replied:
> Now when rethinking this decision it might be best to use the old value to ensure compatibility.

I think there's an decent argument for that. We're also now requiring that an extra executable path be configured properly when one would do before.

As I said in other comments here, I think it's a little awkward you can set `--apache-bin` but it may never be used, but it's not so bad. Unless there's a good reason to switch to the Apache binary that I haven't heard yet, I think I'd like us to:

1. Switch back to using `ctl` here.
2. Don't check for the path to `bin` unless it's going to be used.

joohoi commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -269,8 +293,13 @@ def _check_aug_version(self):
     def get_parser(self):
         """Initializes the ApacheParser"""
+        # If user provided vhost_root value in command line, use it
+        if self.option("vhost_root") != self.OS_DEFAULTS["vhost_root"]:
```
All of the default `vhost-root` paths are parsed (with the exception of Ubuntu/Debian sites-available/sites-enabled scheme)  by the root Apache configuration. The only situation where this would be causing issues, would be if user has modified Apache root configuration to explicitly not parse the default VirtualHost path, but still wanting to create the VirtualHosts to that directory path. I would be surprised if stumbling upon such scenario. Then again, the experience has shown that there's a lot ot crazy Apache configurations out there in the wild.

bmw replied:
Yeah. What do you think about the suggestion of switching back to `self.conf` for the vhost root to preserve behavior here until this is fixed?

bmw commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -97,48 +98,77 @@ class ApacheConfigurator(augeas_configurator.AugeasConfigurator):
+                "ctl", "bin"]
+        for o in opts:
+            # Config options use dashes instead of underscores
+            if self.conf(o.replace("_", "-")) != None:
```
nit: We should use `is not None` here. See the 2nd (and 3rd) bullet points at https://www.python.org/dev/peps/pep-0008/#programming-recommendations.

bmw commented:
I wasn't sure if this was ready for a 2nd review yet because not all of my previous review comments have been addressed, but I gave it one anyway.

joohoi commented:
This is now ready for another round of review. Things that are fixed / changed:
    - I did confirm that both Gentoo and macOS can handle version queries through the ctl script!
    - Additionally also -t -D DUMP_RUN_CFG and -t -D DUMP_MODULES etc work perfectly using ctl script, so I was able to remove "apache_cmd" from the constants as well, reducing complexity!
    - self.conf("vhost-root") is now checked for user defined values, and used even if they are the OS_DEFAULT values
    - Fixed alt_restart_cmd value for CentOS override

With these changes I think all of the review comments were also addressed.

While testing custom vhost paths, I stumbled upon a bug that I'll fix in a separate PR: If a user defined `--apache-vhost-root` is not parsed by Apache configuration, the challenge records are correctly written to the `VirtualHost` but it's still not parsed by apache. We need to include that to the root `httpd.conf` / `apache2.conf` as well.

bmw commented on certbot-apache/certbot_apache/configurator.py:
```diff
@@ -269,8 +293,13 @@ def _check_aug_version(self):
+            vhostroot = None
         return parser.ApacheParser(
-            self.aug, self.conf("server-root"), self.conf("vhost-root"),
+            self.aug, self.option("server_root"), vhostroot,
```
nit: Won't just passing `self.conf("vhost-root")` here get the same behavior as the `if` statements above?

joohoi replied:
The `server_root` is a required parameter to parser, while vhost root is not. This is changed in order to make constant / cli variable handling more streamlined for overrides and main class alike. If we ever need to add functionality around setting or determining `server_root`, this allows it to be done in `ApacheConfigurator._prepare_options()`. For this reason I'd like to keep it as it is.

bmw replied:
I don't understand. What does the change I proposed have to do with `server_root`?

joohoi replied:
Oh, I'm sorry, I misread the comment. You are absolutely correct here. This is now fixed.
